backends/p4tools/modules/testgen/core/small_step/expr_stepper.h

```diff
@@ -18,6 +17,149 @@ namespace P4Tools::P4Testgen {
 
 /// Implements small-step operational semantics for expressions.
 class ExprStepper : public AbstractStepper {
+    /**************************************************************************************************
+    ExternMethodImpls
+    **************************************************************************************************/
+ public:
+    /// Encapsulates a set of extern method implementations.
+    template <typename StepperType>
+    class ExternMethodImpls {
+     public:
+        /// The type of an extern-method implementation. @param externInfo is general useful
+        /// information that can be consumsed by externs (arguments, method names, etc). @param
+        /// stepper is the abstract stepper invoking the function. Templated on the type of stepper
+        /// invoking it.
+        using MethodImpl = std::function<void(const ExternInfo &externInfo, StepperType &stepper)>;
+
+        /// @param call the original method call expression, can be used for stepInto calls.
+        /// @param receiver a symbolic value representing the object on which the method is being
+        ///     called.
+        /// @param name the name of the method being called.
+        /// @param args the list of arguments being passed to method.
+        /// @param state the state in which the call is being made, with the call at the top of the
+        ///     current continuation body.
+        ///
+        /// @returns true if a matching extern-method implementation was found; false otherwise.
+        ///
+        /// A BUG occurs if the receiver is not an extern or if the call is ambiguous.
+        std::optional<MethodImpl> find(const IR::PathExpression &externObjectRef,
+                                       const IR::ID &methodName,
+                                       const IR::Vector<IR::Argument> &args) const {
+            // We have to check the extern type here. We may receive a specialized canonical type,
+            // which we need to unpack.
+            const IR::Type_Extern *externType = nullptr;
+            if (const auto *type = externObjectRef.type->to<IR::Type_Extern>()) {
+                externType = type;
+            } else if (const auto *specType =
+                           externObjectRef.type->to<IR::Type_SpecializedCanonical>()) {
+                CHECK_NULL(specType->substituted);
+                externType = specType->substituted->checkedTo<IR::Type_Extern>();
+            } else if (externObjectRef.path->name == IR::ID("*method")) {
+            } else {
+                BUG("Not a valid extern: %1% with member %2%. Type is %3%.", externObjectRef,
+                    methodName, externObjectRef.type->node_type_name());
+            }
+
+            cstring qualifiedMethodName = externType->name + "." + methodName;
+            if (impls.count(qualifiedMethodName) == 0) {
+                return std::nullopt;
+            }
+
+            const auto &submap = impls.at(qualifiedMethodName);
```

Comment:
Using `it = imps.find` + `it != impls.end()` pattern is preferable to double lookup.